### ISSUE TITLE
Completions fail when gopkgs is not an executable

### ISSUE THREAD

**Issue by thekvs** [Microsoft/vscode-go#2220]
I am using Ubuntu and manually compile all necessary third-party go tools  (`gocode`, `gocode-gomod` etc) and install them in `/usr/local/bin` folder via self-made `.deb` package. And though `/usr/local/bin`  is in `$PATH` it seems that vscode unable to use **some** of that tools like `gocode` and **is** able to use others like `gometalinter`. Everything works when vscode compiles **all** tools itself and places them into `$GOPATH/bin` but for some reason I'd like to maintain aforementioned go tools in separate manner because I also want to use them from the another editor that I occasionally use.

**Comment by ramya-rao-a** [comment]
If you are using modules, then you need to get `gocode` from https://github.com/stamblerre/gocode and rename the resulting binary to `gocode-gomod`. Follow this up with also installing `gocode` from https://github.com/mdempsky/gocode.

Let me know if that helps.

**Comment by thekvs** [comment]
No, all `gocode`s' versions were correct. But it seems that I've found what was causing problems with `vscode-go` plugin --  I've used a wrong path for `gopkgs` tool in my build script so it wasn't built as an executable. Maybe you should somehow address this situation in the extension but that's another issue so I am closing this one.

**Comment by ramya-rao-a** [comment]
> I've used a wrong path for gopkgs tool in my build script so it wasn't built as an executable. Maybe you should somehow address this situation in the extension but that's another issue so I am closing this one.

Can you elaborate on this?

**Comment by thekvs** [comment]
```
kvs@lair ~:> go get -u -d github.com/uudashr/gopkgs    <--- I was using this wrong path in my build script
kvs@lair ~:> go build -i -o /tmp/gopkgs -ldflags="-s -w" github.com/uudashr/gopkgs 
kvs@lair ~:> ls -l /tmp/gopkgs
-rw-rw-r-- 1 kvs kvs 66666 дек 31 09:12 /tmp/gopkgs
kvs@lair ~:> file /tmp/gopkgs 
/tmp/gopkgs: current ar archive   <--- Not an executable!
kvs@lair ~:> rm /tmp/gopkgs
kvs@lair ~:> go build -i -o /tmp/gopkgs -ldflags="-s -w" github.com/uudashr/gopkgs/cmd/gopkgs  <--- This is the correct one
kvs@lair ~:> ls -l /tmp/gopkgs
-rwxrwxr-x 1 kvs kvs 3838272 дек 31 09:12 /tmp/gopkgs
```

**Comment by ramya-rao-a** [comment]
> Maybe you should somehow address this situation in the extension

By this you mean that if the extension does find say a `gopkgs` file as in your case, it should check if it is an executable or not?

**Comment by thekvs** [comment]
Yes.

And the other thing that I'd propose is not to break the whole code completion things (but warn user) if there are execution problems with some auxiliary tools involved in code completion. What I mean is that when I did `chmod a+x` on the wrongly compiled `gopkgs` file (it didn't make it a correct executable!) code completion magically started to work, probably with some reduced functionality.

**Comment by ramya-rao-a** [comment]
We already check whether the file corresponding to the Go tools exist or not. It is a simple step from there to add another check on whether it is an executable as well.

PRs are most welcome to fix this.

Code Pointers:
- [Build and Debug extension](https://github.com/Microsoft/vscode-go/wiki/Building,-Debugging-and-Sideloading-the-extension-in-Visual-Studio-Code#building-and-debugging-the-extension)
- Update the [fileExists](https://github.com/Microsoft/vscode-go/blob/master/src/goPath.ts#L89) function check to do the additional check if the file is an executable. We already have access to the [fs.stats](https://nodejs.org/api/fs.html#fs_class_fs_stats) object at this point, so checking the `mode` should do the trick.

**Comment by keremgocen** [comment]
I'd like to jump on this if no one else have yet

**Comment by ramya-rao-a** [comment]
Go ahead @keremgocen :)

**Comment by oneslash** [comment]
@ramya-rao-a, shall we close this issue since the PR has been already merged?
